File: lib/domain/denpa_men/denpa_men_resistance_calculator.dart
import '../master_data/master_data.dart';
import 'abnormality_resistance.dart';
import 'attribute_resistance.dart';
import 'attribute_resistance_calculator.dart';
import 'color_abnormality_resistance_calculator.dart';
import 'denpa_men.dart';

/// The abnormality/attribute resistance a [DenpaMen] ends up with, as
/// produced by [DenpaMenResistanceCalculation.calculateResistances].
typedef DenpaMenResistances = ({
  List<AbnormalityResistance> abnormalityResistances,
  List<AttributeResistance> attributeResistance,
});

/// Derives a [DenpaMen]'s resistances in a fixed pipeline:
/// 1. Read the body colors and apply their abnormality resistance, then
///    their attribute resistance.
/// 2. Read the head shape and layer its abnormality resistance, then its
///    attribute resistance, on top.
/// 3. Return the combined result.
extension DenpaMenResistanceCalculation on DenpaMen {
  DenpaMenResistances calculateResistances(MasterData masterData) {
    final colorSelection = (bodyColors: bodyColors, isSpColor: isSpColor);

    final abnormalityTotals = <String, int>{
      for (final resistance
          in colorSelection.calculateColorAbnormalityResistance(masterData))
        resistance.abnormalityId: resistance.value,
    };
    final attributeTotals = <String, int>{
      for (final resistance
          in colorSelection.calculateAttributeResistance(masterData))
        resistance.attributeId: resistance.value,
    };

    headShape.abnormalityResistanceBonuses.forEach((abnormalityId, bonus) {
      abnormalityTotals[abnormalityId] =
          (abnormalityTotals[abnormalityId] ?? 0) + bonus;
    });
    headShape.attributeResistanceBonuses.forEach((attributeId, bonus) {
      attributeTotals[attributeId] = (attributeTotals[attributeId] ?? 0) + bonus;
    });

    final attributeIndexById = {
      for (final attribute in masterData.attributes)
        attribute.id: attribute.index,
    };

    return (
      abnormalityResistances: [
        for (final entry in abnormalityTotals.entries)
          if (entry.value != 0)
            AbnormalityResistance(
              abnormalityId: entry.key,
              value: entry.value,
            ),
      ],
      attributeResistance:
          [
            for (final entry in attributeTotals.entries)
              if (entry.value != 0)
                AttributeResistance(attributeId: entry.key, value: entry.value),
          ]..sort(
            (a, b) => (attributeIndexById[a.attributeId] ?? 0).compareTo(
              attributeIndexById[b.attributeId] ?? 0,
            ),
          ),
    );
  }
}

File: modules/server/src/domain/resistance/denpa-men-resistance-calculator.ts
import { calculateAttributeResistance } from './attribute-resistance-calculator';
import { calculateColorAbnormalityResistance } from './color-abnormality-resistance-calculator';
import type {
  BodyColorSelection,
  DenpaMenResistances,
  HeadShapeResistanceBonuses,
  ResistanceMasterData,
} from './types';

/**
 * Ports `DenpaMenResistanceCalculation.calculateResistances` from
 * `lib/domain/denpa_men/denpa_men_resistance_calculator.dart`. Unlike the
 * Dart original (an extension method on `DenpaMen`), this takes the body
 * color selection and head shape bonuses as plain arguments, since the
 * server has no `DenpaMen`-equivalent entity — individual data stays
 * client-side in ObjectBox.
 */
export function calculateDenpaMenResistances(
  colorSelection: BodyColorSelection,
  headShape: HeadShapeResistanceBonuses,
  masterData: ResistanceMasterData,
): DenpaMenResistances {
  const abnormalityTotals = new Map<string, number>();
  for (const resistance of calculateColorAbnormalityResistance(colorSelection, masterData)) {
    abnormalityTotals.set(resistance.abnormalityId, resistance.value);
  }
  const attributeTotals = new Map<string, number>();
  for (const resistance of calculateAttributeResistance(colorSelection, masterData)) {
    attributeTotals.set(resistance.attributeId, resistance.value);
  }

  for (const [abnormalityId, bonus] of Object.entries(headShape.abnormalityResistanceBonuses)) {
    abnormalityTotals.set(abnormalityId, (abnormalityTotals.get(abnormalityId) ?? 0) + bonus);
  }
  for (const bonus of headShape.attributeResistanceBonuses) {
    attributeTotals.set(bonus.attributeId, (attributeTotals.get(bonus.attributeId) ?? 0) + bonus.bonus);
  }

  const attributeIndexById = new Map(masterData.attributes.map((attribute) => [attribute.id, attribute.index]));

  const abnormalityResistances = [...abnormalityTotals.entries()]
    .filter(([, value]) => value !== 0)
    .map(([abnormalityId, value]) => ({ abnormalityId, value }));

  const attributeResistance = [...attributeTotals.entries()]
    .filter(([, value]) => value !== 0)
    .map(([attributeId, value]) => ({ attributeId, value }))
    .sort(
      (a, b) => (attributeIndexById.get(a.attributeId) ?? 0) - (attributeIndexById.get(b.attributeId) ?? 0),
    );

  return { abnormalityResistances, attributeResistance };
}
